### PR TITLE
Refactor PMLogger.java to javascript model.

### PR DESCRIPTION
**Pull request by jlhughes**
- Create Interface for PMLogger
- Move pm cron and services setup to package journal files.
- made PM.combine public - I call it while creating PMs for cluster hops
- set logger to stdoutlogger - so no need to test for null
- add a return;   on recursion.

### CODE REVIEW

**Review by KolombetM — APPROVED** [review]

File: src/foam/nanos/boot/NSpecFactory.java [hunk 1]
```diff
@@ -27,42 +28,52 @@ public NSpecFactory(ProxyX x, NSpec spec) {
...
-      if ( logger != null ) logger.info("Creating Service", spec_.getName());
+      logger.info("Creating Service", spec_.getName());
       ns_ = spec_.createService(x_.getX().put(NSpec.class, spec_));
       Object ns = ns_;
       while ( ns != null ) {
-        if ( ns instanceof ContextAware ) ((ContextAware) ns).setX(x_.getX());
-        if ( ns instanceof NSpecAware )   ((NSpecAware) ns).setNSpec(spec_);
-        if ( ns instanceof NanoService )  ((NanoService) ns).start();
+        if ( ns instanceof ContextAware ) {
+          // REVIEW: System is presently dependent on this blind setX call. - Joel
```
Comment (kgrgreer) [inline src/foam/nanos/boot/NSpecFactory.java:54]: what do you want to review?

File: src/foam/nanos/pm/PM.js [hunk 2]
```diff
@@ -84,7 +84,9 @@ foam.CLASS({
       javaCode: `
     if ( x == null ) return;
     if ( getIsError() ) return;
-    setEndTime(new java.util.Date());
+    if ( getEndTime() == null ) {
+      setEndTime(new java.util.Date());
```
Comment (kgrgreer) [inline src/foam/nanos/pm/PM.js:88]: This reminds me that using a Date is much more expensive that using a long timestamp in millieseconds, BTW.
